isis/tests/Area3DTests.cpp

```diff
@@ -0,0 +1,281 @@
+  Isis::Area3D area2(area1);
+  EXPECT_TRUE(area2.isValid());
+  EXPECT_TRUE(area2.getStartX() == area1.getStartX());
+  EXPECT_TRUE(area2.getStartY() == area1.getStartY());
+  EXPECT_TRUE(area2.getStartZ() == area1.getStartZ());
+  EXPECT_TRUE(area2.getEndX() == area1.getEndX());
+  EXPECT_TRUE(area2.getEndY() == area1.getEndY());
+  EXPECT_TRUE(area2.getEndZ() == area1.getEndZ());
+  EXPECT_TRUE(area2.getHeight() == area1.getHeight());
+  EXPECT_TRUE(area2.getWidth() == area1.getWidth());
+  EXPECT_TRUE(area2.getDepth() == area1.getDepth());
+}
+
+TEST(Area3D, InvalidInputConstructor)
+{
+  Isis::Displacement d1(0, Isis::Displacement::Meters);
+  Isis::Distance d2(0, Isis::Distance::Meters);
+  Isis::Area3D area1(Isis::Displacement(), d1, d1, d1, d1, d1);
+  Isis::Area3D area2(d1, d1, d1, d2, Isis::Distance(), d2);
+  EXPECT_FALSE(area1.isValid());
+  EXPECT_FALSE(area2.isValid());
+}
+
+TEST(Area3D, EndPointCalculations)
+{
+  Isis::Displacement x(10, Isis::Displacement::Meters);
+  Isis::Displacement y(-15, Isis::Displacement::Meters);
+  Isis::Displacement z(20, Isis::Displacement::Meters);
+  Isis::Distance width(10, Isis::Distance::Meters);
+  Isis::Distance height(5, Isis::Distance::Meters);
+  Isis::Distance depth(5, Isis::Distance::Meters);
+  Isis::Area3D area(x, y, z, width, height, depth);
+  EXPECT_DOUBLE_EQ(area.getEndX().meters(), 20);
+  EXPECT_DOUBLE_EQ(area.getEndY().meters(), -10);
+  EXPECT_DOUBLE_EQ(area.getEndZ().meters(), 25);
+}
+
+TEST(Area3D, DimensionCalculations)
+{
+  Isis::Displacement x0(-10, Isis::Displacement::Meters);
+  Isis::Displacement y0(0, Isis::Displacement::Meters);
+  Isis::Displacement z0(-1, Isis::Displacement::Meters);
+  Isis::Displacement x1(50, Isis::Displacement::Meters);
+  Isis::Displacement y1(25, Isis::Displacement::Meters);
+  Isis::Displacement z1(99, Isis::Displacement::Meters);
+  Isis::Area3D area(x0, y0, z0, x1, y1, z1);
+  EXPECT_DOUBLE_EQ(area.getWidth().meters(), 60);
+  EXPECT_DOUBLE_EQ(area.getHeight().meters(), 25);
+  EXPECT_DOUBLE_EQ(area.getDepth().meters(), 100);
+}
+
+TEST(Area3D, Intersect)
+{
+  Isis::Displacement start1(0, Isis::Displacement::Meters);
+  Isis::Distance dim1(1, Isis::Distance::Meters);
+  Isis::Displacement start2(0, Isis::Displacement::Meters);
+  Isis::Distance dim2(0.5, Isis::Distance::Meters);
+  Isis::Area3D area1(start1, start1, start1, dim1, dim1, dim1);
+  Isis::Area3D area2(start2, start2, start2, dim2, dim2, dim2);
+  EXPECT_TRUE(area1.intersect(area2) == area2);
+
+  dim2.setMeters(1.5);
+  area2.setWidth(dim2);
+  area2.setHeight(dim2);
+  area2.setDepth(dim2);
+  EXPECT_TRUE(area1.intersect(area2) == area1);
+
+  start2.setMeters(0.5);
+  area2.setStartX(start2);
+  area2.setStartY(start2);
+  area2.setStartZ(start2);
+  dim1.setMeters(0.5);
+  Isis::Area3D area3(start2, start2, start2, dim1, dim1, dim1);
+  EXPECT_TRUE(area1.intersect(area2) == area3);
+  EXPECT_DOUBLE_EQ(area3.getEndX().meters(), 1);
```

Comment:
Is there a reason to check the endx? Should we be checking the other coordinates then?

Reply:
I was implementing something wrong so I added the getEndX check to help me figure out what was going wrong and I must have forgotten to remove it. It has been removed now.